### PR TITLE
In test-kitchen, install chocolatey behind proxy

### PR DESCRIPTION
**Pull request by dougireton**
I know this is a major refactor and will require a major version bump. I didn't change the cookbook version in the metadata so if you merge this you will need to bump the version before releasing to Supermarket.

It took quite a bit of work to make everything work in test-kitchen behind a proxy. I also added full unit test coverage for the install recipe, and post-convergence integration tests.

* Refactored install script (and .kitchen.yml) to support installing Chocolatey
in test-kitchen behind a proxy.
* Remove `node['chocolatey']['Uri']` attribute. Instead, install by downloading
the chocolatey.nupkg directly via Chef's `remote_file` resource. This lets users
install via Chef client.rb proxy settings instead of inheriting Internet Explorer
proxy settings.

* Add ChefSpec unit tests
* Add ServerSpec integration tests
* Gemfile: bump foodcritic to 5.0 and Berkshelf to 4.0

### CODE REVIEW

**Comment by dougireton** [comment]
One more small bug. Closing temporarily

**Comment by mwrock** [comment]
I'll take a look at where you are at shortly. You might find [this](https://github.com/chef/chef/pull/4106) interesting. I don't think it affects this at all since you are not touching the resource/provider.

**Comment by dougireton** [comment]
Ok. This should be good to go.

**Comment by dougireton** [comment]
@mwrock let me know if you want me to rebase this on top of current master.

File: libraries/helpers.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+module Chocolatey
+  module Helpers
+    # Get the ChocolateyInstall directory from the environment.
+    def chocolatey_install
```
Comment (mwrock) [inline libraries/helpers.rb]: The previous behavior would get the machine value but fall back to the user if it exists and the machine was missing in order not to break older chocolatey clients.

File: libraries/helpers.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+      ENV.fetch('ChocolateyInstall') { |env_var| machine_env_var(env_var) }
+    end
+
+    # The Chocolatey command.
+    #
+    # Reference: https://github.com/chocolatey/chocolatey-cookbook/pull/16#issuecomment-47975896
+    def chocolatey_executable
```
Comment (mwrock) [inline libraries/helpers.rb]: adding the '.exe' could break older chocolatey clients.

Comment (mwrock) [inline libraries/helpers.rb]: Maybe we just need better comments here to explain why we are not using choco.exe and using the odd environment variable checking.

File: recipes/default.rb [hunk 3]
```diff
@@ -17,45 +17,42 @@
...
-file 'cygwin log' do
-  path 'C:/cygwin/var/log/setup.log'
-  action :delete
+powershell_script 'Install Chocolatey' do
+  action :nothing
+  cwd File.join(extract_dir, 'tools')
+  code install_ps1
 end
 
-chocolatey 'chocolatey' do
```
Comment (mwrock) [inline recipes/default.rb]: we don't want to lose the ability to force a chocolatey client upgrade.

**Comment by mwrock** [comment]
Overall I am very much in favor of the installation strategy here which makes better sense for chef clients.  I also really like the overall "modernization" edits. I still want to do some testing but I have added some things I think need tweaking above.

@ferventcoder do you see any potential issue bypassing `install.ps1` and going straight to the `.nupkg`?

**Comment by mwrock** [comment]
@dougireton a rebase would be super

File: libraries/helpers.rb (hunk 1 above)
Comment (dougireton) [inline libraries/helpers.rb]: @mwrock the new behavior tries the user env var first then falls back to fetching the Machine env var. This seems more robust than the previous helper method:

```ruby
# previous method
  def self.chocolatey_install
    ci_keys = ENV.keys.grep(/^ChocolateyInstall$/i)
    ci_keys.count > 0 ? ENV[ci_keys.first] : nil
  end
```

With my new method I was trying to cover the case where chocolatey has just been installed but the user ENV vars haven't been reloaded yet.

File: libraries/helpers.rb (hunk 2 above)
Comment (dougireton) [inline libraries/helpers.rb]: Would you like me to remove the `.exe`?

File: recipes/default.rb (hunk 3 above)
Comment (dougireton) [inline recipes/default.rb]: Agreed, but pretty sure the way I wrote it, the default recipe will fetch based on comparing e-tags from https://chocolatey.org the latest chocolatey.nupkg, thus making upgrade redundant. We should chat about this.

File: libraries/helpers.rb (hunk 2 above)
Comment (dougireton) [inline libraries/helpers.rb]: Aren't you already breaking older chocolatey clients by using `chocolatey list`?

```ruby
  # choco /? returns an exit status of -1 with chocolatey 0.9.9 => use list
      cmd = Mixlib::ShellOut.new("#{chocolatey_executable} list -l chocolatey")
```

File: libraries/helpers.rb (hunk 1 above)
Comment (mwrock) [inline libraries/helpers.rb]: Ok now that I look at the former function again it makes no sense because I don't understand why `ENV` would return multiple keys even if they are mixed case. However, the os should be doing the fallback for us in both cases. If both user and machine `ChocolateyInstall` env vars are set, windows should return the user and fall back to the machine. I remember you mentioning to me offline about trouble with the environment variables. Did you have issues prompting you to force the fallback?

File: libraries/helpers.rb (hunk 2 above)
Comment (mwrock) [inline libraries/helpers.rb]: yes please

Comment (mwrock) [inline libraries/helpers.rb]: no. the older chocolatey created a choco.bat shim

**Comment by dougireton** [comment]
OK, I rebased on top of master and re-pushed

File: recipes/default.rb (hunk 3 above)
Comment (mwrock) [inline recipes/default.rb]: ahh. right. so we should not install if chocolatey is already installed UNLESS `upgrade` is forced. This was the previous behavior which seems reasonable. Ideally, one could set a version attribute to pin to a specific version of chocolatey. For some reason I thought this cookbook had that but I don't see it now. That could be a separate pr.

Thaoughts?

Comment (dougireton) [inline recipes/default.rb]: One way to accomplish this would be to add a not_if block to the remote_file:

```ruby
remote_file nuget_package_path do
  source node['chocolatey']['nupkg']['url']
  backup false
  notifies :unzip, "windows_zipfile[#{extract_dir}]", :immediately
  notifies :run, 'powershell_script[Install Chocolatey]', :immediately
  notifies :run, 'ruby_block[Ensure chocolatey.nupkg is in chocolatey/lib/chocolatey/]', :immediately
  not_if { chocolatey_installed? && (node['chocolatey']['upgrade'] == false) }
end
```

Comment (mwrock) [inline recipes/default.rb]: yeah. I think that would work.

**Comment by ferventcoder** [comment]
> do you see any potential issue bypassing install.ps1 and going straight to the .nupkg?

Not really, but you might want to take a look at the latest changes now, which support a proxy - https://chocolatey.org/install.ps1

**Comment by ferventcoder** [comment]
~~~powershell
# Environment Variables, specified as $env:NAME in PowerShell.exe and %NAME% in cmd.exe.
# For explicit proxy, please set $env:chocolateyProxyLocation and optionally $env:chocolateyProxyUser and $env:chocolateyProxyPassword
# For an explicit version of Chocolatey, please set $env:chocolateyVersion = 'versionnumber'
# To target a different url for chocolatey.nupkg, please set $env:chocolateyDownloadUrl = 'full url to nupkg file' 
# NOTE: $env:chocolateyDownloadUrl does not work with $env:chocolateyVersion.
# To use built-in compression (no 7zip download), please set $env:chocolateyUseWindowsCompression = 'true'
~~~

**Comment by ferventcoder** [comment]
My preference is that folks have their own internal server and they put the chocolatey version they want there. Then they would use the URI to install from there. Very similar to https://github.com/chocolatey/puppet-chocolatey#use-an-internal-chocolateynupkg-for-chocolatey-installation and https://github.com/chocolatey/puppet-chocolatey/blob/master/templates/InstallChocolatey.ps1.erb

File: spec/matcher_spec.rb [hunk 4]
```diff
@@ -1,18 +1,20 @@
 require 'chefspec'
 require 'chefspec/berkshelf'
 
-describe 'chocolatey_specs::default' do
+RSpec.describe 'chocolatey_specs::default' do
```
Comment (mwrock) [inline spec/matcher_spec.rb]: why preface with `RSpec`?

Comment (mwrock) [inline spec/matcher_spec.rb]: never mind. I see this is the "new" way of doing things and it makes sense. Just wasn't used to seeing it.

**Comment by mwrock** [comment]
This is awesome @dougireton . Its all working well here.

I think in light of some recent changes @ferventcoder made to the `install.ps1` we can further tweak the `default` recipe to make this a non breaking change and still work with proxies.

See his comment above about the new proxy environment variables. We can set those and then still use the `install.ps1` to work with proxies. However instead of using `powershell_out!` to pull down that script via a .net web client, we'll use a `remote_file` to grab it.

I'd love to release this today so I'm gonna make this change now. I'll ping you later today to see if you can test on your end behind a proxy before I release. If you are not available then no worries. We can do it next week.
